### ISSUE TITLE
Prepare slower than Appender

### ISSUE THREAD

**Issue by TheGiamig** [duckdb/duckdb#1336]
Hi, I'm trying to speed up a data recorder/replayer based on SQLite.

I coded an application to transfer data from old DB to DuckDB.
As first attempt I used the connection.Prepared + Execute methods to insert data.
The tranfert speed wasn't that high..
Also using the C-api the transfert seems really slow.
Then I re-wrote the thing using connection.Query obtaining much better performances.

There is something I completely missed or this beaviour is correct?
Thanks

**Comment by TheGiamig** [comment]
Sorry, I wrote connection.Query but I was talking about the Appender.

**Comment by Mytherin** [comment]
This is fully expected; the appender is optimized for exactly this use case. Prepare avoids e.g. re-running the optimization and planning phase but still needs to do more than just the Appender. For example, it sets up the query context and creates a result object for every query that is run. There is likely room for improving that somewhat, but using the appender is preferable since the code there is much simpler and faster. 

Another thing you should do is wrap multiple statements in a `BEGIN TRANSACTION` and `COMMIT` block. This also applies to the appender. If you do not do this but run in auto commit mode, there will be the additional overhead of starting and committing transactions, which includes syncing data to disk if you are not running in in-memory mode. This is extremely slow. i.e. your code should look like this:

```cpp
con.BeginTransaction();
for(...) {
   con.Execute(1, 2, 3, 4);
}
con.Commit();
```

Or like this with the appender:

```cpp
con.BeginTransaction();
for(...) {
	appender.AppendRow(1, 2, 3, 4);
}
con.Commit();
```

**Comment by TheGiamig** [comment]
Thank you Mark!
Yeah I know transactions will remove some overhead (same for SQLite, huge speed improvement).
Have you some tips for me to speed up the reading? Transaction is a nop in this case?
Thanks

**Comment by Mytherin** [comment]
That is hard to answer without more details about what you are doing/trying to do/what API you are using. Perhaps you could expand a bit on your problem, the speed you are encountering and the speed you expect?

**Comment by TheGiamig** [comment]
Oh I mean something not so specific, just some basic tips as Appender or transactions but for the read side..
..but since you insist.. :)
.. I wrote a simple read test using the old SQL queries. The improvement compared to SQLite is about 20X.
Very impressive.
The heavy part of my DB is a sort of timed series (key, blob, timestamp). I got millions of these tuple in a recording session.
When replaying the data, I need to retrieve the last blob for a given (key, time) or retrieve all tuples in a given time range.
The thing is not exactly this way but is just to explain a little more.

So, back to my test, I executed a read banchmark:
- a loop calling Execute -> about 21sec
- a loop calling Query   -> about 6sec
- SQLite                         -> 120sec 

The query for the Prepare is `"SELECT type, value, MAX(time) FROM attribute WHERE object_id = $1 AND time <= $2 GROUP BY type, value;"`.

The Query call instead is that
`auto result = con->Query("SELECT type, value, MAX(time) FROM attribute WHERE object_id = " + std::to_string(object_id) + " AND time <= " + std::to_string(time) + " GROUP BY type, value;");`

I don't understand why the Prepare + Execute is slower than the simple Query call.
